### PR TITLE
Various reified extensions on NamedDomainObjectContainerExtensions

### LINKED ISSUE #605: NamedDomainObjectContainerExtensions should have reified versions of API's
### Expected Behavior

You should be able to do things like:
```kotlin
val test by tasks.getting<Test> { useTestNG() }
```

Currently, the API only supports using the `NamedDomainObjectContainerExtensions` with API's requiring the `Klass` type.
So currently you have to do this:
```kotlin
val test by tasks.getting(Test::class) { useTestNG() }
```

### Context
The requirement to pass the `Klass` type is very un-kotliny. The reified type reads better and also allows for this sort of syntax:

```kotlin
val test : Test by tasks.getting { useTestNG() }
```

### Your Environment
Current API: https://github.com/gradle/kotlin-dsl/blob/6a3c51a65e8423aa33432b816996360cd32e7906/provider/src/main/kotlin/org/gradle/kotlin/dsl/NamedDomainObjectContainerExtensions.kt

### PR DESCRIPTION
**Pull request by JLLeitschuh**
Closes #605

### Contributor Checklist
- [x] Base the PR against the `develop` branch
- [x] [Sign Gradle CLA](http://gradle.org/contributor-license-agreement/)
- [ ] Provide integration tests to verify changes from a user perspective
- [x] Provide unit tests to verify logic
- [ ] Ensure that tests pass locally: `./gradlew check --parallel`

### CODE REVIEW

File: provider/src/main/kotlin/org/gradle/kotlin/dsl/NamedDomainObjectContainerExtensions.kt [hunk 1]
```diff
@@ -79,24 +95,54 @@ class NamedDomainObjectContainerScope<T : Any>(
...
     fun <U : T> getting(type: KClass<U>) =
         polymorphicDomainObjectContainer().getting(type)
 
+    /**
+     * Provides a property delegate that gets elements of the given [U].
+     */
+    inline
+    fun <reified U : T> getting() =
+        getting(U::class)
+
```
Comment (bamboo) [inline provider/src/main/kotlin/org/gradle/kotlin/dsl/NamedDomainObjectContainerExtensions.kt]: :x: Unnecessary empty line.

File: provider/src/test/kotlin/org/gradle/kotlin/dsl/NamedDomainObjectCollectionExtensionsTest.kt [hunk 2]
```diff
@@ -110,6 +112,21 @@ class NamedDomainObjectCollectionExtensionsTest {
...
 
+    @Test
+    fun `can access existing element by getting with reified type`() {
+        val element = DomainObject()
+        val container = mock<PolymorphicDomainObjectContainer<SomeObject>> {
+            on { getByName("domainObject") } doReturn element
+        }
+
+        container { // invoke syntax reified method
+            val domainObject by getting<DomainObject>()
```
Comment (bamboo) [inline provider/src/test/kotlin/org/gradle/kotlin/dsl/NamedDomainObjectCollectionExtensionsTest.kt]: :o: is `val domainObject: DomainObject by getting` also allowed?

File: provider/src/test/kotlin/org/gradle/kotlin/dsl/NamedDomainObjectCollectionExtensionsTest.kt [hunk 3]
```diff
@@ -110,6 +112,21 @@ class NamedDomainObjectCollectionExtensionsTest {
...
+            assertThat(domainObject, sameInstance(element))
+        }
+
+        // regular syntax reified method won't have the same behaviour or compile time guarantees
```
Comment (bamboo) [inline provider/src/test/kotlin/org/gradle/kotlin/dsl/NamedDomainObjectCollectionExtensionsTest.kt]: :o: This means reified syntax is not available for qualified collection references such as `val myTask by tasks.getting<MyTask>` which creates an asymmetry in the DSL that IIRC was the reason we didn't include it in the first round.

We also have to solve the related problem of not providing the _invoke syntax DSL_ when plugins declare configuration methods such as `plugins(configure: Action<PluginContainer>)`. We should consider this PR in light of that. 

@eskatos any thoughts?

File: provider/src/main/kotlin/org/gradle/kotlin/dsl/NamedDomainObjectContainerExtensions.kt [hunk 4]
```diff
@@ -79,24 +95,54 @@ class NamedDomainObjectContainerScope<T : Any>(
...
+    fun getting(configuration: T.() -> Unit) =
+        polymorphicDomainObjectContainer().getting(configuration)
+
     /**
      * Provides a property delegate that gets elements of the given [type] and applies the given [configuration].
      */
     fun <U : T> getting(type: KClass<U>, configuration: U.() -> Unit) =
         polymorphicDomainObjectContainer().getting(type, configuration)
 
+    inline
```
Comment (eskatos) [inline provider/src/main/kotlin/org/gradle/kotlin/dsl/NamedDomainObjectContainerExtensions.kt]: :x: Missing kdoc

File: provider/src/test/kotlin/org/gradle/kotlin/dsl/NamedDomainObjectCollectionExtensionsTest.kt (hunk 2 above)
Comment (JLLeitschuh) [reply provider/src/test/kotlin/org/gradle/kotlin/dsl/NamedDomainObjectCollectionExtensionsTest.kt]: Yes, there is a test above that covers that.

File: provider/src/test/kotlin/org/gradle/kotlin/dsl/NamedDomainObjectCollectionExtensionsTest.kt (hunk 3 above)
Comment (JLLeitschuh) [reply provider/src/test/kotlin/org/gradle/kotlin/dsl/NamedDomainObjectCollectionExtensionsTest.kt]: Would you like me to remove the reified `getting` methods?

File: provider/src/main/kotlin/org/gradle/kotlin/dsl/NamedDomainObjectContainerExtensions.kt (hunk 1 above)
Comment (JLLeitschuh) [reply provider/src/main/kotlin/org/gradle/kotlin/dsl/NamedDomainObjectContainerExtensions.kt]: This should be resolved now. Would really love a linter to check this for me.

**Comment by JLLeitschuh** [comment]
I think I resolved your major concerns. Not sure if either of you have any additional thoughts on my changes to `getting`. I'm not really attached to those changes and would be fine with removing them if that would be better.

File: provider/src/main/kotlin/org/gradle/kotlin/dsl/NamedDomainObjectContainerExtensions.kt [hunk 5]
```diff
@@ -79,24 +95,56 @@ class NamedDomainObjectContainerScope<T : Any>(
     fun <U : T> creating(type: KClass<U>) =
         polymorphicDomainObjectContainer().creating(type)
 
+    /**
+     * Provides a property delegate that creates elements of the given [U].
+     */
+    inline fun <reified U : T> creating() =
```
Comment (bamboo) [inline provider/src/main/kotlin/org/gradle/kotlin/dsl/NamedDomainObjectContainerExtensions.kt]: :x: `inline` modifier goes in the line above.

File: provider/src/main/kotlin/org/gradle/kotlin/dsl/NamedDomainObjectContainerExtensions.kt [hunk 6]
```diff
@@ -79,24 +95,56 @@ class NamedDomainObjectContainerScope<T : Any>(
...
     /**
      * Provides a property delegate that creates elements of the given [type] with the given [configuration].
      */
     fun <U : T> creating(type: KClass<U>, configuration: U.() -> Unit) =
         polymorphicDomainObjectContainer().creating(type, configuration)
 
+    /**
+     * Provides a property delegate that creates elements of the given [U] with the given [configuration].
+     */
+    inline fun <reified U : T> creating(noinline configuration: U.() -> Unit) =
```
Comment (bamboo) [inline provider/src/main/kotlin/org/gradle/kotlin/dsl/NamedDomainObjectContainerExtensions.kt]: :x: `inline` modifier goes in the line above.

File: provider/src/main/kotlin/org/gradle/kotlin/dsl/NamedDomainObjectContainerExtensions.kt (hunk 5 above)
Comment (JLLeitschuh) [reply provider/src/main/kotlin/org/gradle/kotlin/dsl/NamedDomainObjectContainerExtensions.kt]: Dang, I accidentally ran the ktlint formatter I think, oops.

**Comment by bamboo** [comment]
Hey @JLLeitschuh, is `./gradlew check --parallel` passing for you?

**Comment by JLLeitschuh** [comment]
I've had issues with running out of file descriptors when running the whole thing in parallel.

Let me try this PR again and see what happens.

**Comment by eskatos** [comment]
Closing as the solution shouldn't introduce any asymmetry in the DSL and will be a bit different.

**Pull request closed without merging**